src/VisualStudio/CSharp/Impl/CSharpPackage.cs

```diff
@@ -64,22 +65,24 @@ namespace Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService
 
     [ProvideAutomationProperties("TextEditor", "CSharp", Guids.TextManagerPackageString, profileNodeLabelId: 101, profileNodeDescriptionId: 106, resourcePackageGuid: Guids.CSharpPackageIdString)]
     [ProvideAutomationProperties("TextEditor", "CSharp-Specific", packageGuid: Guids.CSharpPackageIdString, profileNodeLabelId: 104, profileNodeDescriptionId: 105)]
-    [ProvideService(typeof(CSharpLanguageService), ServiceName = "C# Language Service")]
-    [ProvideService(typeof(ICSharpTempPECompilerService), ServiceName = "C# TempPE Compiler Service")]
+    [ProvideService(typeof(CSharpLanguageService), ServiceName = "C# Language Service", IsAsyncQueryable = true)]
+    [ProvideService(typeof(ICSharpTempPECompilerService), ServiceName = "C# TempPE Compiler Service", IsAsyncQueryable = true)]
     internal class CSharpPackage : AbstractPackage<CSharpPackage, CSharpLanguageService>, IVsUserSettingsQuery
     {
         private ObjectBrowserLibraryManager _libraryManager;
         private uint _libraryManagerCookie;
 
-        protected override void Initialize()
+        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
             try
             {
-                base.Initialize();
+                await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
+
+                await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
                 this.RegisterService<ICSharpTempPECompilerService>(() => new TempPECompilerService(this.Workspace.Services.GetService<IMetadataService>()));
 
-                RegisterObjectBrowserLibraryManager();
+                await RegisterObjectBrowserLibraryManagerAsync(cancellationToken).ConfigureAwait(true);
             }
             catch (Exception e) when (FatalError.Report(e))
```

Comment:
Needs change to ReportUnlessCanceled() now that cancellation is expected? Seems bad!

Reply:
What is the scenario for the cancellation token being canceled here? Is it user shutting down VS during an async package load? 

---
In reply to: [187726239](https://github.com/dotnet/roslyn/pull/26787#discussion_r187726239) [](ancestors = 187726239)

Reply:
That's right: `GetServiceAsync` doesn't take a `CancellationToken` parameter. The only way a package or service init gets canceled is on shutdown, AFAIK.

Reply:
BTW, if a service you query for is canceled because of shutdown, your code will get a null service back. Yet another reason why folks should always check for null after a GetService call and do reasonable stuff that doesn't involve crashing the process. :)

Reply:
@AArnott To confirm your statement: calling GetService once shutdown has begun could return null, even prior to AsyncPackage. Is there an additional thing now that a package initialize being cancelled can return null while in the middle of the initialize?

Also, are there any test harness to randomly trigger cancellation part of the way in on a service fetch? Beause first rule of software: if it's not tested, it's broken. No matter how reasonable we try to be, we won't succeed.

Reply:
Hmm. Sounds like we may want to make a helper for GetService that is cancellation aware. 

---
In reply to: [187741403](https://github.com/dotnet/roslyn/pull/26787#discussion_r187741403) [](ancestors = 187741403)

Reply:
> Is there an additional thing now that a package initialize being cancelled can return null while in the middle of the initialize?

I'm not sure I understand your question. Package Initialize doesn't return anything (well, besides Task itself, which had better not be null). Being an async package yourself at this point, you might start to initialize and end up throwing OperationCanceledException instead of completing, which you'd never have done as a sync Package. Your *callers* then are more likely to see null when they query for services your package proffers.

Reply:
@jaredpar What would the helper method do differently? Throw instead of return null? We do have such a method now, on [`IAsyncServiceProvider2`](http://index/?leftProject=Microsoft.VisualStudio.Shell.Framework&leftSymbol=iewyn4cvk4mm&file=IAsyncServiceProvider2.cs&rightSymbol=fnyjjunoh8wk).

Reply:
@AArnott Let me rephrase then: if I am in InitializeAsync(), and call GetService() to get _another_ package's service, can that return null due to cancellation I haven't observed yet?

Reply:
(obviously pre-AsyncPackage world where *everybody* was synchronous, that would only give null if the service couldn't be loaded.)

Reply:
> Now if once you get the null result, if you then check your own cancellationToken

That was the basic idea I had for the wrapper. 

> I don't know whether you can count on that being canceled by that point. @ryanmolden may be able to answer that.

Feel like we need that guarantee else how can we reliably program against this model? If services which can't return `null` do so and then we need a way to discover this is expected because we are in shutdown / cancellation.

Reply:
Hmm. If the token is guaranteed to be cancelled, then I'd wonder for Roslyn's codebase at least if we should ban GetServiceAsync(Type), and create GetServiceAsync(Type, CancellationToken). If it returns null, then recheck the token and convert that to a cancellation throw. That way the cancellation path only has to be handled in one consistent way (via exception) and the "we actually failed to get the service" can be handled in a consistent way too.

Reply:
(I suspect that's what @jaredpar means about a cancellation aware GetServiceAsync)

Reply:
Yep.

Reply:
>  If services which can't return null do so and then we need a way to discover this is expected because we are in shutdown / cancellation.

Querying for *any* service might return null. It might be VS safe mode, it might be shutdown, or that a service/package didn't get installed. Or it might simply be a failure within the service.

My own recipe is that if you can gracefully proceed without a service, do it. If you can't, then throw. No need to crash the process. Just throw. I use `Assumes.Present` for this since it throws with a good exception message that explains which service was expected but was null, automatically. That effectively propagates the null you got to your caller. And ultimately, packages/services fail which will be communicated to the user -- or quietly swallowed when in shutdown.

Reply:
@jasonmalinowski For your `GetServiceAsync` method proposal, I think that's what `IAsyncServiceProvider2` effectively does: it throws the same exception that the service you queried for threw. That turns out to be great because if you allow that exception to propagate (since you need the service and can't operate without it), then that means that ultimately the activity log, the user, and/or the crash will all point at the root cause rather than blame you, as the innocent middle-man.

Reply:
> My own recipe is that if you can gracefully proceed without a service, do it. If you can't, then throw.

Agree. Hence my suggestion that we wrap this into a tidy little helper.

Reply:
> it throws the same exception that the service you queried for threw

The docs for IAsyncServiceProvider2 say:

> It does not guarantee exceptions will be thrown however for failure cases.

and has a lot of other hedging in it's statements. If that is what it does do, that documentation needs heavy rewording to ensure the contract. :smile:

Still, the case that @jaredpar and I worry about is even if we call IAsyncServiceProvider2.GetServiceAsync(), and we're in the cancellation case, can we get null if the other proffering package was never asked? If *that* package threw cancellation then yes it may give the cancellation exception. But can it decide early that it gets null, and that's OK?

Reply:
>  If that is what it does do, that documentation needs heavy rewording to ensure the contract.

Sadly, I think the docs are correct. We added that method in order to allow exceptions to propagate, which it does, but in some cases (perhaps the one you speculated, I don't know) it can still return null. Maybe when the service isn't registered and the package hasn't proffered it, such that the service simply doesn't exist. 

> Hence my suggestion that we wrap this into a tidy little helper.

Here's what you can do today:

```cs
IAsyncServiceProvider2 asp;
var service = await asp.GetServiceAsync(typeof(SVsService), swallowExceptions: false) as IVsService;
Assumes.Present(service);
```

Is it the last two lines that you'd like to see simplified to something like this?

```
IAsyncServiceProvider2 asp;
var service = await asp.GetServiceOrThrowAsync(typeof(SVsService)) as IVsService;
```

Reply:
It's the cancellation handling that is worrisome to me. I want to be able to put prior to your Assumes.Present() a cancellationToken.ThrowIfCancellationRequested(), and have the guarantee that it'll actually work.

Reply:
(and then wrap *that* in a helper)

Reply:
Maybe we can get `AsyncPackage.GetServiceAsync` to do the first two lines (query, and throw cancellation if it's null), and then add a `AsyncPackage.GetServiceOrThrowAsync<TInterface>(Type)` method that guarantees a null result will never come back by calling the first method then throwing if null.

Reply:
so, new GetServiceAsync will have cancellation token passed in?

Reply:
➡️ I'm going to update the code to use `cancellationToken.ThrowIfCancellationRequested()` and assume it does the right thing.